File: recursion/merge-two-sorted-lists.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} list1
# @param {ListNode} list2
# @return {ListNode}
def merge_two_lists(list1, list2)
    dummy = ListNode.new("dummy")
    curr = dummy
    while list1 && list2
        if list1.val < list2.val
            curr.next = list1
            list1 = list1.next
        else
            curr.next = list2
            list2 = list2.next
        end
        curr = curr.next
    end
    if list1
        curr.next = list1
    end
    if list2
        curr.next = list2
    end
    dummy.next
end

File: recursion/merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def mergeTwoLists(self, list1, list2):
        """
        :type list1: Optional[ListNode]
        :type list2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        dummy = ListNode(-1)
        curr = dummy
        while list1 and list2:
            if list1.val < list2.val:
                curr.next = ListNode(list1.val)
                list1 = list1.next  
            else:
                curr.next = ListNode(list2.val)
                list2 = list2.next  
            curr = curr.next
        if list1:
            curr.next = list1
        if list2:
            curr.next = list2
        return dummy.next